File: usecase/vfs_usecase.go
package usecase

import (
	"context"
	"fmt"
	"log"

	"github.com/gotd/td/tg"
	"github.com/teledrive/teledrive/domain"
	tginfra "github.com/teledrive/teledrive/infrastructure/telegram"
)

// VFSUsecase handles virtual file system operations including folder/file CRUD and trash.
type VFSUsecase struct {
	folderRepo domain.FolderRepository
	fileRepo   domain.FileRepository
	clientPool *tginfra.ClientPool
}

// NewVFSUsecase creates a new VFSUsecase.
func NewVFSUsecase(folderRepo domain.FolderRepository, fileRepo domain.FileRepository, clientPool *tginfra.ClientPool) *VFSUsecase {
	return &VFSUsecase{
		folderRepo: folderRepo,
		fileRepo:   fileRepo,
		clientPool: clientPool,
	}
}

// DirectoryListing contains the folders and files in a directory.
type DirectoryListing struct {
	Folders []domain.Folder `json:"folders"`
	Files   []domain.File   `json:"files"`
}

// ListDirectory returns all non-deleted folders and files under the given parent.
func (uc *VFSUsecase) ListDirectory(userID int, folderID *int) (*DirectoryListing, error) {
	folders, err := uc.folderRepo.ListByParent(userID, folderID)
	if err != nil {
		return nil, fmt.Errorf("failed to list folders: %w", err)
	}

	files, err := uc.fileRepo.ListByFolder(userID, folderID)
	if err != nil {
		return nil, fmt.Errorf("failed to list files: %w", err)
	}

	return &DirectoryListing{
		Folders: folders,
		Files:   files,
	}, nil
}

// CreateFolder creates a new folder under the given parent.
func (uc *VFSUsecase) CreateFolder(userID int, name string, parentID *int) (*domain.Folder, error) {
	if name == "" {
		return nil, fmt.Errorf("folder name is required")
	}

	// If parentID is provided, verify it exists and belongs to the user.
	if parentID != nil {
		parent, err := uc.folderRepo.GetByID(*parentID)
		if err != nil {
			return nil, fmt.Errorf("parent folder not found: %w", err)
		}
		if parent.UserID != userID {
			return nil, fmt.Errorf("parent folder does not belong to user")
		}
	}

	folder := &domain.Folder{
		UserID:   userID,
		Name:     name,
		ParentID: parentID,
	}

	created, err := uc.folderRepo.Create(folder)
	if err != nil {
		return nil, fmt.Errorf("failed to create folder: %w", err)
	}
	return created, nil
}

// RenameFolder renames a folder.
func (uc *VFSUsecase) RenameFolder(userID int, folderID int, newName string) (*domain.Folder, error) {
	if newName == "" {
		return nil, fmt.Errorf("new name is required")
	}

	folder, err := uc.folderRepo.GetByID(folderID)
	if err != nil {
		return nil, fmt.Errorf("folder not found: %w", err)
	}
	if folder.UserID != userID {
		return nil, fmt.Errorf("folder does not belong to user")
	}

	folder.Name = newName
	if err := uc.folderRepo.Update(folder); err != nil {
		return nil, fmt.Errorf("failed to rename folder: %w", err)
	}
	return folder, nil
}

// RenameFile renames a file.
func (uc *VFSUsecase) RenameFile(userID int, fileID int, newName string) (*domain.File, error) {
	if newName == "" {
		return nil, fmt.Errorf("new name is required")
	}

	file, err := uc.fileRepo.GetByID(fileID)
	if err != nil {
		return nil, fmt.Errorf("file not found: %w", err)
	}
	if file.UserID != userID {
		return nil, fmt.Errorf("file does not belong to user")
	}

	file.Name = newName
	if err := uc.fileRepo.Update(file); err != nil {
		return nil, fmt.Errorf("failed to rename file: %w", err)
	}
	return file, nil
}

// MoveFolder moves a folder to a new parent.
func (uc *VFSUsecase) MoveFolder(userID int, folderID int, newParentID *int) (*domain.Folder, error) {
	folder, err := uc.folderRepo.GetByID(folderID)
	if err != nil {
		return nil, fmt.Errorf("folder not found: %w", err)
	}
	if folder.UserID != userID {
		return nil, fmt.Errorf("folder does not belong to user")
	}

	// Prevent moving a folder into itself.
	if newParentID != nil && *newParentID == folderID {
		return nil, fmt.Errorf("cannot move folder into itself")
	}

	// If newParentID is provided, verify it exists and belongs to the user.
	if newParentID != nil {
		parent, err := uc.folderRepo.GetByID(*newParentID)
		if err != nil {
			return nil, fmt.Errorf("target parent folder not found: %w", err)
		}
		if parent.UserID != userID {
			return nil, fmt.Errorf("target parent folder does not belong to user")
		}
	}

	folder.ParentID = newParentID
	if err := uc.folderRepo.Update(folder); err != nil {
		return nil, fmt.Errorf("failed to move folder: %w", err)
	}
	return folder, nil
}

// MoveFile moves a file to a new folder.
func (uc *VFSUsecase) MoveFile(userID int, fileID int, newFolderID *int) (*domain.File, error) {
	file, err := uc.fileRepo.GetByID(fileID)
	if err != nil {
		return nil, fmt.Errorf("file not found: %w", err)
	}
	if file.UserID != userID {
		return nil, fmt.Errorf("file does not belong to user")
	}

	// If newFolderID is provided, verify it exists and belongs to the user.
	if newFolderID != nil {
		folder, err := uc.folderRepo.GetByID(*newFolderID)
		if err != nil {
			return nil, fmt.Errorf("target folder not found: %w", err)
		}
		if folder.UserID != userID {
			return nil, fmt.Errorf("target folder does not belong to user")
		}
	}

	file.FolderID = newFolderID
	if err := uc.fileRepo.Update(file); err != nil {
		return nil, fmt.Errorf("failed to move file: %w", err)
	}
	return file, nil
}

// DeleteFolder soft-deletes a folder.
func (uc *VFSUsecase) DeleteFolder(userID int, folderID int) error {
	folder, err := uc.folderRepo.GetByID(folderID)
	if err != nil {
		return fmt.Errorf("folder not found: %w", err)
	}
	if folder.UserID != userID {
		return fmt.Errorf("folder does not belong to user")
	}

	if err := uc.folderRepo.SoftDelete(folderID); err != nil {
		return fmt.Errorf("failed to delete folder: %w", err)
	}
	return nil
}

// DeleteFile soft-deletes a file.
func (uc *VFSUsecase) DeleteFile(userID int, fileID int) error {
	file, err := uc.fileRepo.GetByID(fileID)
	if err != nil {
		return fmt.Errorf("file not found: %w", err)
	}
	if file.UserID != userID {
		return fmt.Errorf("file does not belong to user")
	}

	if err := uc.fileRepo.SoftDelete(fileID); err != nil {
		return fmt.Errorf("failed to delete file: %w", err)
	}
	return nil
}

// TrashListing contains soft-deleted folders and files.
type TrashListing struct {
	Folders []domain.Folder `json:"folders"`
	Files   []domain.File   `json:"files"`
}

// ListTrash returns all trashed folders and files for a user.
func (uc *VFSUsecase) ListTrash(userID int) (*TrashListing, error) {
	folders, err := uc.folderRepo.ListTrashed(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to list trashed folders: %w", err)
	}

	files, err := uc.fileRepo.ListTrashed(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to list trashed files: %w", err)
	}

	return &TrashListing{
		Folders: folders,
		Files:   files,
	}, nil
}

// RestoreFolder restores a soft-deleted folder.
func (uc *VFSUsecase) RestoreFolder(userID int, folderID int) error {
	folder, err := uc.folderRepo.GetByID(folderID)
	if err != nil {
		return fmt.Errorf("folder not found: %w", err)
	}
	if folder.UserID != userID {
		return fmt.Errorf("folder does not belong to user")
	}

	if err := uc.folderRepo.Restore(folderID); err != nil {
		return fmt.Errorf("failed to restore folder: %w", err)
	}
	return nil
}

// RestoreFile restores a soft-deleted file.
func (uc *VFSUsecase) RestoreFile(userID int, fileID int) error {
	file, err := uc.fileRepo.GetByID(fileID)
	if err != nil {
		return fmt.Errorf("file not found: %w", err)
	}
	if file.UserID != userID {
		return fmt.Errorf("file does not belong to user")
	}

	if err := uc.fileRepo.Restore(fileID); err != nil {
		return fmt.Errorf("failed to restore file: %w", err)
	}
	return nil
}

// EmptyTrash permanently deletes all trashed folders and files for a user.
func (uc *VFSUsecase) EmptyTrash(userID int) error {
	// Retrieve all trashed files first
	trashedFiles, err := uc.fileRepo.ListTrashed(userID)
	if err != nil {
		return fmt.Errorf("failed to list trashed files: %w", err)
	}

	// Group message IDs to delete from Telegram
	var messageIDs []int
	for _, file := range trashedFiles {
		if file.TelegramMessageID != 0 {
			messageIDs = append(messageIDs, file.TelegramMessageID)
		}
	}

	if len(messageIDs) > 0 {
		client, err := uc.clientPool.GetClient(userID, "")
		if err == nil {
			api := client.API()
			// Telegram API allows deleting up to 100 messages at once
			chunkSize := 100
			for i := 0; i < len(messageIDs); i += chunkSize {
				end := i + chunkSize
				if end > len(messageIDs) {
					end = len(messageIDs)
				}
				chunk := messageIDs[i:end]
				_, err = api.MessagesDeleteMessages(context.Background(), &tg.MessagesDeleteMessagesRequest{
					Revoke: true, // Delete for everyone
					ID:     chunk,
				})
				if err != nil {
					log.Printf("Failed to delete messages %v from Telegram: %v", chunk, err)
				}
			}
		} else {
			log.Printf("Failed to get Telegram client for user %d to delete messages: %v", userID, err)
		}
	}

	// Delete all trashed files from the database
	if err := uc.fileRepo.PermanentDeleteAllTrashed(userID); err != nil {
		return fmt.Errorf("failed to empty trashed files: %w", err)
	}

	// Delete all trashed folders.
	trashedFolders, err := uc.folderRepo.ListTrashed(userID)
	if err != nil {
		return fmt.Errorf("failed to list trashed folders: %w", err)
	}
	for _, folder := range trashedFolders {
		if err := uc.folderRepo.PermanentDelete(folder.ID); err != nil {
			return fmt.Errorf("failed to permanently delete folder %d: %w", folder.ID, err)
		}
	}

	return nil
}

// GetBreadcrumb returns the breadcrumb trail from root to the given folder.
func (uc *VFSUsecase) GetBreadcrumb(userID int, folderID int) ([]domain.BreadcrumbItem, error) {
	items, err := uc.folderRepo.GetBreadcrumb(folderID, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get breadcrumb: %w", err)
	}
	return items, nil
}

// ListStarred returns all non-deleted starred folders and files for a user.
func (uc *VFSUsecase) ListStarred(userID int) (*DirectoryListing, error) {
	folders, err := uc.folderRepo.ListStarred(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to list starred folders: %w", err)
	}

	files, err := uc.fileRepo.ListStarred(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to list starred files: %w", err)
	}

	return &DirectoryListing{
		Folders: folders,
		Files:   files,
	}, nil
}



// ToggleStar toggles the starred status of a folder or file.
func (uc *VFSUsecase) ToggleStar(userID int, itemType string, itemID int) error {
	if itemType == "folder" {
		folder, err := uc.folderRepo.GetByID(itemID)
		if err != nil {
			return fmt.Errorf("folder not found: %w", err)
		}
		if folder.UserID != userID {
			return fmt.Errorf("permission denied")
		}
		folder.IsStarred = !folder.IsStarred
		return uc.folderRepo.Update(folder)
	} else if itemType == "file" {
		file, err := uc.fileRepo.GetByID(itemID)
		if err != nil {
			return fmt.Errorf("file not found: %w", err)
		}
		if file.UserID != userID {
			return fmt.Errorf("permission denied")
		}
		file.IsStarred = !file.IsStarred
		return uc.fileRepo.Update(file)
	}
	return fmt.Errorf("invalid item type: %s", itemType)
}

File: src/usecase/vfs.usecase.ts
import { Api } from 'telegram';
import type { Folder, FolderRepository, BreadcrumbItem } from '../domain/folder.js';
import type { File, FileRepository } from '../domain/file.js';
import type { ClientPool } from '../infrastructure/telegram/client-pool.js';

export interface DirectoryListing {
  folders: Folder[];
  files: File[];
}

export class VFSUsecase {
  private folderRepo: FolderRepository;
  private fileRepo: FileRepository;
  private clientPool: ClientPool;

  constructor(
    folderRepo: FolderRepository,
    fileRepo: FileRepository,
    clientPool: ClientPool
  ) {
    this.folderRepo = folderRepo;
    this.fileRepo = fileRepo;
    this.clientPool = clientPool;
  }

  public async listDirectory(userId: number, folderId: number | null): Promise<DirectoryListing> {
    const folders = await this.folderRepo.listByParent(userId, folderId);
    const files = await this.fileRepo.listByFolder(userId, folderId);

    return {
      folders,
      files,
    };
  }

  public async createFolder(userId: number, name: string, parentId: number | null): Promise<Folder> {
    if (!name || !name.trim()) {
      throw new Error('Folder name is required');
    }

    if (parentId !== null) {
      const parent = await this.folderRepo.getById(parentId);
      if (!parent) {
        throw new Error('Parent folder not found');
      }
      if (parent.user_id !== userId) {
        throw new Error('Parent folder does not belong to user');
      }
    }

    return await this.folderRepo.create({
      user_id: userId,
      name: name.trim(),
      parent_id: parentId,
    });
  }

  public async renameFolder(userId: number, folderId: number, newName: string): Promise<Folder> {
    if (!newName || !newName.trim()) {
      throw new Error('New name is required');
    }

    const folder = await this.folderRepo.getById(folderId);
    if (!folder) {
      throw new Error('Folder not found');
    }
    if (folder.user_id !== userId) {
      throw new Error('Folder does not belong to user');
    }

    folder.name = newName.trim();
    await this.folderRepo.update({
      id: folder.id,
      name: folder.name,
      parent_id: folder.parent_id,
      is_starred: folder.is_starred,
    });

    return folder;
  }

  public async renameFile(userId: number, fileId: number, newName: string): Promise<File> {
    if (!newName || !newName.trim()) {
      throw new Error('New name is required');
    }

    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }

    file.name = newName.trim();
    await this.fileRepo.update({
      id: file.id,
      name: file.name,
      folder_id: file.folder_id,
      is_starred: file.is_starred,
    });

    return file;
  }

  public async moveFolder(userId: number, folderId: number, newParentId: number | null): Promise<Folder> {
    const folder = await this.folderRepo.getById(folderId);
    if (!folder) {
      throw new Error('Folder not found');
    }
    if (folder.user_id !== userId) {
      throw new Error('Folder does not belong to user');
    }

    if (newParentId !== null && newParentId === folderId) {
      throw new Error('Cannot move folder into itself');
    }

    if (newParentId !== null) {
      const parent = await this.folderRepo.getById(newParentId);
      if (!parent) {
        throw new Error('Target parent folder not found');
      }
      if (parent.user_id !== userId) {
        throw new Error('Target parent folder does not belong to user');
      }
    }

    folder.parent_id = newParentId;
    await this.folderRepo.update({
      id: folder.id,
      name: folder.name,
      parent_id: folder.parent_id,
      is_starred: folder.is_starred,
    });

    return folder;
  }

  public async moveFile(userId: number, fileId: number, newFolderId: number | null): Promise<File> {
    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }

    if (newFolderId !== null) {
      const folder = await this.folderRepo.getById(newFolderId);
      if (!folder) {
        throw new Error('Target folder not found');
      }
      if (folder.user_id !== userId) {
        throw new Error('Target folder does not belong to user');
      }
    }

    file.folder_id = newFolderId;
    await this.fileRepo.update({
      id: file.id,
      name: file.name,
      folder_id: file.folder_id,
      is_starred: file.is_starred,
    });

    return file;
  }

  public async deleteFolder(userId: number, folderId: number): Promise<void> {
    const folder = await this.folderRepo.getById(folderId);
    if (!folder) {
      throw new Error('Folder not found');
    }
    if (folder.user_id !== userId) {
      throw new Error('Folder does not belong to user');
    }

    try {
      const messageIds = await this.collectFolderFileIds(userId, folderId);
      if (messageIds.length > 0) {
        const client = await this.clientPool.getClient(userId);
        await client.deleteMessages('me', messageIds, { revoke: true });
      }
    } catch (tgErr) {
      console.warn(`Failed to delete messages for folder ${folderId} from Telegram:`, tgErr);
    }

    await this.folderRepo.softDelete(folderId);
  }

  public async deleteFile(userId: number, fileId: number): Promise<void> {
    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }

    // Delete message from Telegram Saved Messages
    if (file.telegram_message_id && file.telegram_message_id > 0) {
      try {
        const client = await this.clientPool.getClient(userId);
        await client.deleteMessages('me', [file.telegram_message_id], { revoke: true });
      } catch (tgErr) {
        console.warn(`Failed to delete message ${file.telegram_message_id} from Telegram:`, tgErr);
      }
    }

    await this.fileRepo.softDelete(fileId);
  }

  private async collectFolderFileIds(userId: number, folderId: number): Promise<number[]> {
    let messageIds: number[] = [];
    const directFiles = await this.fileRepo.listByFolder(userId, folderId);
    for (const f of directFiles) {
      if (f.telegram_message_id > 0) messageIds.push(f.telegram_message_id);
      await this.fileRepo.softDelete(f.id);
    }

    const childFolders = await this.folderRepo.listByParent(userId, folderId);
    for (const child of childFolders) {
      const childMsgIds = await this.collectFolderFileIds(userId, child.id);
      messageIds = messageIds.concat(childMsgIds);
      await this.folderRepo.softDelete(child.id);
    }

    return messageIds;
  }

  public async listTrash(userId: number): Promise<DirectoryListing> {
    const folders = await this.folderRepo.listTrashed(userId);
    const files = await this.fileRepo.listTrashed(userId);

    return {
      folders,
      files,
    };
  }

  public async restoreFolder(userId: number, folderId: number): Promise<void> {
    const folder = await this.folderRepo.getById(folderId);
    if (!folder) {
      throw new Error('Folder not found');
    }
    if (folder.user_id !== userId) {
      throw new Error('Folder does not belong to user');
    }

    await this.folderRepo.restore(folderId);
  }

  public async restoreFile(userId: number, fileId: number): Promise<void> {
    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }

    await this.fileRepo.restore(fileId);
  }

  public async emptyTrash(userId: number): Promise<void> {
    const trashedFiles = await this.fileRepo.listTrashed(userId);
    const messageIds = trashedFiles
      .map((f) => f.telegram_message_id)
      .filter((id) => id > 0);

    if (messageIds.length > 0) {
      try {
        const client = await this.clientPool.getClient(userId);
        const chunkSize = 100;
        for (let i = 0; i < messageIds.length; i += chunkSize) {
          const chunk = messageIds.slice(i, i + chunkSize);
          try {
            await client.invoke(
              new Api.messages.DeleteMessages({
                id: chunk,
                revoke: true,
              })
            );
          } catch (delErr) {
            console.warn(`Failed to delete messages ${chunk} from Telegram:`, delErr);
          }
        }
      } catch (clientErr) {
        console.warn(`Failed to get Telegram client to delete messages for user ${userId}:`, clientErr);
      }
    }

    await this.fileRepo.permanentDeleteAllTrashed(userId);

    const trashedFolders = await this.folderRepo.listTrashed(userId);
    for (const folder of trashedFolders) {
      await this.folderRepo.permanentDelete(folder.id);
    }
  }

  public async getBreadcrumb(userId: number, folderId: number): Promise<BreadcrumbItem[]> {
    return await this.folderRepo.getBreadcrumb(folderId, userId);
  }

  public async listStarred(userId: number): Promise<DirectoryListing> {
    const folders = await this.folderRepo.listStarred(userId);
    const files = await this.fileRepo.listStarred(userId);

    return {
      folders,
      files,
    };
  }

  public async toggleStar(userId: number, itemType: 'folder' | 'file', itemId: number): Promise<void> {
    if (itemType === 'folder') {
      const folder = await this.folderRepo.getById(itemId);
      if (!folder) {
        throw new Error('Folder not found');
      }
      if (folder.user_id !== userId) {
        throw new Error('Permission denied');
      }

      await this.folderRepo.update({
        id: folder.id,
        name: folder.name,
        parent_id: folder.parent_id,
        is_starred: !folder.is_starred,
      });
    } else if (itemType === 'file') {
      const file = await this.fileRepo.getById(itemId);
      if (!file) {
        throw new Error('File not found');
      }
      if (file.user_id !== userId) {
        throw new Error('Permission denied');
      }

      await this.fileRepo.update({
        id: file.id,
        name: file.name,
        folder_id: file.folder_id,
        is_starred: !file.is_starred,
      });
    } else {
      throw new Error(`Invalid item type: ${itemType}`);
    }
  }

  public async updateCaption(userId: number, fileId: number, caption: string): Promise<File> {
    const file = await this.fileRepo.getById(fileId);
    if (!file) {
      throw new Error('File not found');
    }
    if (file.user_id !== userId) {
      throw new Error('File does not belong to user');
    }

    await this.fileRepo.updateCaption(fileId, caption);
    file.caption = caption;
    return file;
  }

  public async listAllFiles(userId: number): Promise<File[]> {
    return await this.fileRepo.listAll(userId);
  }

  public async syncWithTelegram(userId: number): Promise<{ added: number; synced: number }> {
    const client = await this.clientPool.getClient(userId);
    const messages = await client.getMessages('me', { limit: 50 });
    const existingFiles = await this.fileRepo.listAll(userId);
    const existingMsgIds = new Set(existingFiles.map((f) => f.telegram_message_id));

    let added = 0;

    for (const msg of messages) {
      if (!msg || !msg.media || existingMsgIds.has(msg.id)) continue;

      let name = '';
      let size = 0;
      let mimeType = 'application/octet-stream';
      let fileId = msg.id.toString();
      const caption = msg.message || '';

      const media: any = msg.media;

      if (media.document || media.className === 'MessageMediaDocument') {
        const doc = media.document || media;
        size = Number(doc.size || 0);
        mimeType = doc.mimeType || 'application/octet-stream';
        fileId = doc.id ? doc.id.toString() : msg.id.toString();

        const nameAttr = doc.attributes?.find(
          (a: any) => a.className === 'DocumentAttributeFilename' || a.fileName
        );
        name = nameAttr?.fileName || `file_${msg.id}`;
      } else if (media.photo || media.className === 'MessageMediaPhoto') {
        const photo = media.photo || media;
        fileId = photo.id ? photo.id.toString() : msg.id.toString();
        mimeType = 'image/jpeg';
        name = `photo_${msg.id}.jpg`;
        const largest = photo.sizes?.[photo.sizes.length - 1];
        size = largest?.size || 1024 * 200;
      } else {
        continue;
      }

      let chatId = '0';
      if (msg.peerId) {
        const peer: any = msg.peerId;
        if (peer.userId) chatId = peer.userId.toString();
        else if (peer.channelId) chatId = peer.channelId.toString();
        else if (peer.chatId) chatId = peer.chatId.toString();
      }

      await this.fileRepo.create({
        user_id: userId,
        folder_id: null,
        name,
        size,
        mime_type: mimeType,
        telegram_message_id: msg.id,
        telegram_chat_id: chatId,
        telegram_file_id: fileId,
        caption,
      });

      added++;
    }

    return {
      added,
      synced: existingFiles.length + added,
    };
  }
}
